### PR TITLE
A faster flash attention bwd implementation

### PR DESCRIPTION
**Pull request by tonywu95**
- Decompose the bwd kernel into two kernels, one for dq and one for dk,dv. 
- Extra parallelism over the sequence length axis.
- On a benchmark, with causal=True, it is close to 6X faster compared to the previous implementation. ~3X faster than XLA bwd pass.

### CODE REVIEW

**Review by sharadmv — CHANGES_REQUESTED** [review]
High level comment: the current backward pass is a fully fused kernel that parallelizes over batch * num heads number of threads.

For attention shapes that have small batch and heads (as is common in language model training) this kernel will underutilize the GPU.

However, there are applications where this kernel is faster than the two kernel variant.

Could you add the two kernel version as a separate backward pass impl, that way the user has the option of selecting the one they want?

File: jax_triton/pallas/ops/attention.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def body(start_k, carry):
                                              "interpret", "debug"])
 def mha(q, k, v,
         sm_scale: float = 1.0,
-        causal: bool = False,
+        causal: bool = True,
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:102]: Could we keep the default as False (for now)?

File: jax_triton/pallas/ops/attention.py [hunk 2]
```diff
@@ -292,6 +292,109 @@ def inner_loop(start_q, carry):
...
+        ds = ds * sm_scale
+      dq = dq + pl.dot(ds.astype(k.dtype), k).astype(dq.dtype)
+      return dq
+  if causal:
+      upper_bound = lax.div(start_q * block_q, block_k) + 1
+  else:
+      upper_bound = jt.cdiv(seq_len, block_k)
+  dq = lax.fori_loop(0, upper_bound, inner_loop, dq)
+  pl.store(dq_ref, (pl.ds(start_q * block_q, block_q),
+                        slice(None)), dq, eviction_policy="evict_last")
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:343]: Style nit: fix indentation

File: jax_triton/pallas/ops/attention.py [hunk 3]
```diff
@@ -292,6 +292,109 @@ def inner_loop(start_q, carry):
...
+  start_k = pl.program_id(0)
+
+  dv = jnp.zeros([block_k, block_d], dtype=jnp.float32)
+  dk = jnp.zeros([block_k, block_d], dtype=jnp.float32)
+  k = pl.load(k_ref, (pl.ds(start_k * block_k, block_k), slice(None)))
+  v = pl.load(v_ref, (pl.ds(start_k * block_k, block_k), slice(None)))
+  span_k = start_k * block_k + jnp.arange(block_k)
+
+  def inner_loop(start_q, carry):
+      dv, dk = carry
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:365]: Style nit: we use two space indents in this repo. Could you fix here and elsewhere?

File: jax_triton/pallas/ops/attention.py [hunk 4]
```diff
@@ -292,6 +292,109 @@ def inner_loop(start_q, carry):
...
+  seq_len = q_ref.shape[0]
+
+  start_q = pl.program_id(0)
+  q = pl.load(q_ref, (pl.ds(start_q * block_q, block_q), slice(None)))
+  span_q = start_q * block_q + jnp.arange(block_q)
+  m = pl.load(m_ref, (pl.ds(start_q * block_q, block_q),))
+  do = pl.load(do_scaled_ref, (pl.ds(start_q * block_q, block_q), slice(None)))
+  di = pl.load(delta_ref, (pl.ds(start_q * block_q, block_q),))
+  dq = pl.load(dq_ref, (pl.ds(start_q * block_q, block_q),
+                        slice(None)), eviction_policy="evict_last")
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:314]: I'm unsure if we need eviction_policy here. Could you try/profile w/ and w/out it? Here and elsewhere.

File: jax_triton/pallas/ops/attention.py [hunk 5]
```diff
@@ -292,6 +292,109 @@ def inner_loop(start_q, carry):
...
+
+  def inner_loop(start_k, carry):
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:316]: Can just rename `carry ` to `dq` in the arguments.

File: jax_triton/pallas/ops/attention.py [hunk 6]
```diff
@@ -292,6 +292,109 @@ def inner_loop(start_q, carry):
...
+      dq = carry
+      span_k = start_k * block_k + jnp.arange(block_k)
+      k = pl.load(k_ref, (pl.ds(start_k * block_k, block_k), slice(None)))
+      v = pl.load(v_ref, (pl.ds(start_k * block_k, block_k), slice(None)))
+
+      qk = pl.dot(q, k.T)
+      qk = qk.astype(q_ref.dtype)
+      qk = qk.astype(jnp.float32)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:324]: These casts might hurt numerics. Could you remove them to check whether the numerics improve?

File: jax_triton/pallas/ops/attention.py [hunk 7]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
         input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
+  elif backward_pass_impl == "triton_split":
+    # We accumulate into dq so we need to initialize it to zeros.
+    dq = jnp.zeros(q.shape, jnp.float32)
+    out_shapes_q = jax.ShapeDtypeStruct(dq.shape, dq.dtype)
+
+    grid_q = (jt.cdiv(seq_len, block_q), batch_size, num_heads)
+    # grid_q = (batch_size, num_heads)
+    # TODO(sharadmv): figure out why num_warps=8 doesn't work!
+    num_warps = 4
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:460]: Can you try num_warps=8?

File: jax_triton/pallas/ops/attention.py [hunk 8]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
         num_warps=num_warps,
         num_stages=1,
         input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
+  elif backward_pass_impl == "triton_split":
+    # We accumulate into dq so we need to initialize it to zeros.
+    dq = jnp.zeros(q.shape, jnp.float32)
+    out_shapes_q = jax.ShapeDtypeStruct(dq.shape, dq.dtype)
+
+    grid_q = (jt.cdiv(seq_len, block_q), batch_size, num_heads)
+    # grid_q = (batch_size, num_heads)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:458]: Remove comment.

File: jax_triton/pallas/ops/attention.py [hunk 9]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
         num_warps=num_warps,
         num_stages=1,
         input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
+  elif backward_pass_impl == "triton_split":
+    # We accumulate into dq so we need to initialize it to zeros.
+    dq = jnp.zeros(q.shape, jnp.float32)
+    out_shapes_q = jax.ShapeDtypeStruct(dq.shape, dq.dtype)
+
+    grid_q = (jt.cdiv(seq_len, block_q), batch_size, num_heads)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:457]: Could you move the seq_len grid dim to the end (I usually like making my grid kinda look like my input shapes).

File: jax_triton/pallas/ops/attention.py [hunk 10]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        out_specs=[
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        name="mha_backward",
+        debug=debug,
+        interpret=interpret,
+        num_warps=num_warps,
+        num_stages=1,
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:484]: It's also worth trying num_stages=2.

File: jax_triton/pallas/ops/attention.py [hunk 11]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        out_specs=[
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        name="mha_backward",
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:480]: Should be `name="mha_backward_dq"`

File: jax_triton/pallas/ops/attention.py [hunk 12]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+        debug=debug,
+        interpret=interpret,
+        num_warps=num_warps,
+        num_stages=1,
+        input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
+
+    grid_kv = (jt.cdiv(seq_len, block_k), batch_size, num_heads)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:487]: Same here.

File: jax_triton/pallas/ops/attention.py [hunk 13]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+          pl.BlockSpec(lambda _, j, k: (j, k, 0), (None, None, seq_len)),
+        ],
+        out_specs=[
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        name="mha_backward",
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:511]: `name="mha_backward_dkv"`

File: jax_triton/pallas/ops/attention.py [hunk 14]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+        debug=debug,
+        interpret=interpret,
+        num_warps=num_warps,
+        num_stages=1)(q, k, v, out, do_scaled, l, m, delta)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:515]: Also try num_stages=2 here?

File: jax_triton/pallas/ops/attention.py [hunk 15]
```diff
@@ -362,4 +530,4 @@ def mha_reference(q, k, v, sm_scale=1.0, causal: bool = False):
     mask = jnp.broadcast_to(mask, logits.shape)
     logits = jnp.where(mask, logits, float('-inf'))
   weights = jax.nn.softmax(logits * sm_scale).astype(q.dtype)
-  return jnp.einsum('bhqk,bkhc->bqhc', weights, v)
+  return jnp.einsum('bhqk,bkhc->bqhc', weights, v)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:533]: Add newline

**Review by sharadmv — CHANGES_REQUESTED** [review]

File: jax_triton/pallas/ops/attention.py [hunk 16]
```diff
@@ -346,14 +449,79 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+        ],
+        out_specs=[
+          pl.BlockSpec(lambda _, j, k: (j, 0, k, 0), (None, seq_len, None, head_dim)),
+        ],
+        name="mha_backward",
+        debug=debug,
+        interpret=interpret,
+        num_warps=num_warps,
+        num_stages=1,
+        input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:485]: We shouldn't need input_output_aliases for this kernel.

File: jax_triton/pallas/ops/attention.py (hunk 4 above)
Comment (sharadmv) [reply jax_triton/pallas/ops/attention.py:314]: I realize now you can just initialize dq to 0 here. No need for a load.

File: jax_triton/pallas/ops/attention.py (hunk 7 above)
Comment (tonywu95) [reply jax_triton/pallas/ops/attention.py:460]: Yup it works, and it even speeds up the backward pass!

File: jax_triton/pallas/ops/attention.py (hunk 10 above)
Comment (tonywu95) [reply jax_triton/pallas/ops/attention.py:484]: Tried, no improvement but I will leave it as 2.

File: jax_triton/pallas/ops/attention.py (hunk 6 above)
Comment (tonywu95) [reply jax_triton/pallas/ops/attention.py:324]: Tried, no improvement.

File: jax_triton/pallas/ops/attention.py [hunk 17]
```diff
@@ -292,6 +293,109 @@ def inner_loop(start_q, carry):
...
+      ds = ds * sm_scale
+    dq = dq + pl.dot(ds.astype(k.dtype), k).astype(dq.dtype)
+    return dq
+  if causal:
+    upper_bound = lax.div(start_q * block_q, block_k) + 1
+  else:
+    upper_bound = jt.cdiv(seq_len, block_k)
+  dq = lax.fori_loop(0, upper_bound, inner_loop, dq)
+  pl.store(dq_ref, (pl.ds(start_q * block_q, block_q),
+                        slice(None)), dq, eviction_policy="evict_last")
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:342]: I don't think we need eviction policy here

File: jax_triton/pallas/ops/attention.py [hunk 18]
```diff
@@ -292,6 +293,109 @@ def inner_loop(start_q, carry):
...
+  if causal:
+    lower_bound = lax.div(start_k * block_k, block_q)
+  else:
+    lower_bound = 0
+  dv, dk = lax.fori_loop(lower_bound, jt.cdiv(seq_len, block_q), inner_loop,
+                           (dv, dk))
+  pl.store(dv_ref, (pl.ds(start_k * block_k, block_k),
+                      slice(None)), dv.astype(dv_ref.dtype))
+  pl.store(dk_ref, (pl.ds(start_k * block_k, block_k),
+                      slice(None)), dk.astype(dk_ref.dtype))
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:396]: Nit: indentation

File: jax_triton/pallas/ops/attention.py [hunk 19]
```diff
@@ -346,6 +450,65 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
         num_warps=num_warps,
         num_stages=1,
         input_output_aliases={8: 0})(q, k, v, out, do_scaled, l, m, delta, dq)
+  elif backward_pass_impl == "triton_split":
+    # We accumulate into dq so we need to initialize it to zeros.
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:454]: Comment is not accurate here

File: jax_triton/pallas/ops/attention.py [hunk 20]
```diff
@@ -346,6 +450,65 @@ def _mha_backward(sm_scale: float, causal: bool, block_q: int, block_k: int,
...
+    out_shapes_q = jax.ShapeDtypeStruct(q.shape, jnp.float32)
```
Comment (sharadmv) [inline jax_triton/pallas/ops/attention.py:455]: I suspect we don't need dq to be f32 anymore. Could you try q.dtype?

**Review by sharadmv — CHANGES_REQUESTED** [review]
Could you also add tests into pallas_test.py?

**Comment by abhinavgoel95** [comment]
Can this PR be merged? We see a big performance improvement on NVIDIA A100 GPUs with this PR.

**Comment by sharadmv** [comment]
I left some comments. @tonywu95 do you have time to address them?

**Comment by skye** [comment]
Hey @tonywu95, is it ok if we take over this PR and put you as a co-author? We'd love to get it in!

**Comment by superbobry** [comment]
Closing since Pallas Triton kernels are now longer hosted here.
